Added compose to the cloudstorage library

rbruyere opened the pull request:
Added the compose functionality to the cloudstorage library with a basic
test site.

aozarov commented:
I looked at the CL and I have 3 main comments.

1) adhere to style in this file (and be consistent). issues are:
   line width - should be 80
   Use ' instead of "
   Don't use \ (actually, I think, in all cases it was used it was not not necessary [not even the + \])
   No space after colon (e.g. line 424) 
   No space after [ or before ]
2) I think this method is trying to do to much and be too flexible.
   * Though "more powerful" to provide all the feature set such as generation and IfGenerationMatch it
     is not consistent with the other methods (and could be added later if needed).
     Was there a use-case for it? Note that in this case I am OK with having it as is though not my
     top preference (which is just a list of filenames).
   * I don't think you should provide the sorting option (not much value and therefore just clutter the API).
   * I think the "batching" should be done in a higher level (and let them decide on temp name, cleanup and how to compose the intermediate files). I think it is not expected to leave garbage in, lets say, a failure mid way. and operation may take much longer than expected. Operation should accept the limit of a individual compose operation (32).
  3) instead of the internal _make_api_call you should add this functionality to storage_api (like the other methods).

rbruyere commented:
All the lines (according to PyLint) are less than 80.
Fixed the quotes.
Removed '\'
Removed spaces before ':'
Removed extra spaces around  [ ]

I added the Generation because it was available in the XML api that the library was wrapping. I can remove it if you want it doesn't bother me either way.

Adapted to accept list if file names as well. If we drop the generations ill remove the dicts all together.

Removed the sort option.

Pulled batching out into a separate function. Users now have a choice.

Re-factored code to put the call into storage_api. I'll send another pull request shortly.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+def compose(list_of_files, destination_file,
+            content_type=None, retry_params=None, _account_id=None):
+  '''
+    Runs the GCS Compose on the inputed files.
```
s/inputed/inputted/ (I think that is the correct spelling, no?) Though I would prefer "input" or "given"

rbruyere replied:
Fixed

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  '''
+    Runs the GCS Compose on the inputed files.
+    Merges between 2 and 32 files into one file. If any of the previous files had the merge
+    run on them it counts towards the 32 limit.
```
I don't think that is true. the 32 limit is of file names where the 1024 limit is of objects (a filename can be composed by multiple objects). Hence composing a 31 files with another file that is composed from 32 files should be OK (as long as it didn't reach the 1024 limit).

rbruyere replied:
Re-wrote and added link to docs

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    Merges between 2 and 32 files into one file. If any of the previous files had the merge
+    run on them it counts towards the 32 limit.
+  Args:
+    list_of_files: list of file name strings or dictionaries with the following format:
```
Do you think it would have been strange to support list of either string or dictionary?

aozarov replied:
Oh, sorry. I didn't realize that you already added the support of list of strings or dictionaries.
I prefer the optional arguments as suggested above but with with a mix list as well (though not sure how to make that
more clear).

rbruyere replied:
Personally I don't. If you feel its an issue I can split the two, but ti will cause more code needed by both the user an the compose lib.

rbruyere replied:
Response to second comment:
So 2 variables list_of_files (list of strings) and files_metadata (list of dictionaries). Then merge them in code?

aozarov replied:
Yes. that was the idea (list_of_files required, files_metadata optional)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+      {'file_name': REQUIRED name of the file to be merged. Do not include the bucket name,
+       'Generation': OPTIONAL Used to specify what version of a file to use,
+       'IfGenerationMatch': OPTIONAL Used to fail requests if versions don't match}
+    destination_file: Path to the desired output file. Must have the bucket in the path.
```
does it need to start with '/'?

rbruyere replied:
Clarified comment.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+       'Generation': OPTIONAL Used to specify what version of a file to use,
+       'IfGenerationMatch': OPTIONAL Used to fail requests if versions don't match}
+    destination_file: Path to the desired output file. Must have the bucket in the path.
+    content_type: Used to specify the content-header of the output.
```
I don't think you should specify if not provided and instead let GCS pick the default.
Also, should you say "based on" (https://answers.yahoo.com/question/index?qid=20090506215752AAzQrJx)

rbruyere replied:
It defaults to "binary/octet-stream" which is odd imo. I can easily remove the functionality.

aozarov replied:
I agree that this is a wrong choice from GCS (and we should probably point it out to them to pick a better default) but I don't think we should set it if not provided (e.g. GCS could add a global/bucket option for "default mime-type" and we will not pick it up).

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  Raises:
+    TypeError: If the dictionary for the file list is malformed
+    ValueError: If the number of files is outside the range of 2-1024
+    errors.NotFoundError: If any element in the file list is missing the 'file_name' key
```
This sound more like an input error (e.g. ValueError) and not (NotFoundError).

rbruyere replied:
I could make it required like it is in the JSON API. But the XML does not require one and defaults to binary.

aozarov replied:
what defaults to "binary"? I was referring to raising errors.NotFoundError if a file in the file_list does not 'file_name' key.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+            with open(bucket + source_file['file_name'], 'r') as gcs_source:
+              gcs_merge.write(gcs_source.read())
+          except errors.NotFoundError:
+            logging.warn('File not found %s, skipping', source_file['file_name'])
```
Lets create an issue for adding compose to cloudstorage_stub and cleaning up this code when available.
Also, I think you should fail when not found rather than log-and-ignore (to match production behavior)

rbruyere replied:
The file is in the dev server. I don't know if creating an issue here would transfer over.
I need to get this working ASAP to implement it in Mapreduce.

aozarov replied:
The issue should be in this repository to correspond to the TODO.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+# pylint: disable=too-many-locals, too-many-branches, too-many-statements
+def compose(list_of_files, destination_file,
+            content_type=None, retry_params=None, _account_id=None):
+  '''
```
Actually docstring should use """ instead of '''

rbruyere replied:
Fixed

aozarov replied:
I don't see that fix.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+
+def _validate_compose_list(destination_file, file_list, content_type, number_of_files=32):
+  '''
+    Validates the compose list and extracts needed information from it
```
end with '.' and line separator

rbruyere replied:
Fixed

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    except errors.NotFoundError:
+      pass
+
+def file_exists(destination):
```
can this be made "private" as well (_file_exists)?
Also, I think it should be enough just to open the file for read, no?

rbruyere replied:
Why make them private? Having access to the functionality would beneficial to users?
Wasn't sure if an open close (2 api calls) was better than getting a stats. If open close is better I'll change it.

aozarov replied:
Good question. I think open/close would be cheaper.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    If it succeeds returns True otherwise false
+
+    Args:
+      destination: Full path to the file (ie. /bucket/object) w/ leading slash
```
w/ ? with, including,...
Also, we should be consistent about with periods for Args.

rbruyere replied:
replaced w/ with the word with
Changed all args to finish with '.'

aozarov commented on python/test/composet_test.py:
```diff
@@ -0,0 +1,103 @@
+""" cloudstorage compose tester"""
```
should file be called compose_test.py?

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+def file_exists(destination):
+  '''Checks if a file exists.
+    Tries to do a stat check on the file.
+    If it succeeds returns True otherwise false
```
Nit, either change false to False or True to true.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  if list_len <= 1:
+    raise ValueError('Compose operation requires at least two components; %i provided.' % list_len)
+
+  if isinstance(file_list[0], str):
```
If you prefer one list (instead of 2 for files and metadata) then i think we should accept mixed values in the list.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  for source_file in list_of_files:
+    if not isinstance(source_file, dict):
+      raise TypeError('Each item of file_list must be a string or dictionary')
+    file_name = source_file.get('file_name', None)
```
None is already the  default value

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +275,185 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    file_name = source_file.get('file_name', None)
+    if file_name is None:
+      raise errors.NotFoundError('Each item in file_list must specify a file_name')
+    if file_name.startswith(bucket):
```
It should be valid to have a folder name similar to bucket name. should you check that it does not start with '/'?

rbruyere replied:
I agree, that's why it is a warning not an error. Adding check for '/'. Ill do it as a warning as they may have a '/' in the file name (confirmed)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    ValueError: If the number of files is outside the range of 2-32.
+  """
+  api = storage_api._get_storage_api(retry_params=retry_params,
+                                   account_id=_account_id)
```
Fix alignment

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+def compose(list_of_files, destination_file, files_metadata=None,
+            content_type=None, retry_params=None, _account_id=None):
+  """
+    Runs the GCS Compose on the given files.
```
See https://www.python.org/dev/peps/pep-0257/
You should use one of these styles:
1) """Runs the GCS ...
2) """
    Runs the GCS

Though I like the latter the former seems to be consistent with existing docstring in this file.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+            with open(bucket + source_file['Name'], 'r') as gcs_source:
+              gcs_merge.write(gcs_source.read())
+          except errors.NotFoundError as error:
+            logging.warn('File not found %s, skipping', source_file['file_name'])
```
No need for the warning now when you propagate the exception.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    compose_object = _temp_func
+  else:
+    compose_object = api.compose_object
+  file_list, _ = _validate_compose_list(
```
Fix indentation:
file_list, _ = _validate_compose_list(destination_file,
                                                          list_of_files,
                                                          files_metadata, 32)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  Needed until cloudstorage_stub.py is updated to accept compose requests
+  TODO: When patched remove the True flow from this if.
+  """
+  if 'development' in os.environ.get('SERVER_SOFTWARE', '').lower():
```
The "recommended way" to check for dev-appserver is:
 os.getenv('SERVER_SOFTWARE').startswith('Dev')

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+def _compose_batch(list_of_files, destination_file, files_metadata=None,
+            content_type=None, retry_params=None, _account_id=None):
+  """
+  Runs the GCS Compose on the inputed files in batches of 32.
```
I don't think we should have this method. Do you use it?

rbruyere replied:
I do, not only in personal code but the catalyst for me to start this project was to use it in the Mapreduce, which could send up to 256 files for the compose.

aozarov replied:
On Mon, Mar 30, 2015 at 7:25 AM, rbruyere <notifications@github.com> wrote:

> In python/src/cloudstorage/cloudstorage_api.py
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6#discussion_r27394062>
> :
>
> > +          except errors.NotFoundError as error:
> > +            logging.warn('File not found %s, skipping', source_file['file_name'])
> > +            raise error
> > +    compose_object = _temp_func
> > +  else:
> > +    compose_object = api.compose_object
> > +  file_list, _ = _validate_compose_list(
> > +                                      destination_file,
> > +                                      list_of_files,
> > +                                      files_metadata, 32)
> > +  compose_object(file_list, destination_file, content_type)
> > +
> > +def _compose_batch(list_of_files, destination_file, files_metadata=None,
> > +            content_type=None, retry_params=None, _account_id=None):
> > +  """
> > +  Runs the GCS Compose on the inputed files in batches of 32.
>
> I do, not only in personal code but the catalyst for me to start this
> project was to use it in the Mapreduce, which could send up to 256 files
> for the compose.
>
Similar to what we do for Java, such code should be done in the MR (that
could provide the temporary files location and could better handle
cleanup).

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6/files#r27394062>
> .
>

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+      pass
+
+def _file_exists(destination):
+  """Checks if a file exists.
```
method looks good. docstring should change to:
"""Checks if a file exists.
Tries to open the file.
If it succeeds returns True otherwise False.

Args:
  destination: Full path to the file (ie. /bucket/object) with leading slash.

Returns:
  True if the file is accessible otherwise False.
"""

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+      destination: Full path to the file (ie. /bucket/object).
+      file_list: list of files to compose, see compose for details.
+      files_metadata: Meta details for the file
+      number_of_files=32: maximum number of files allowed in the list.
```
number_of_files: Maximum number of files allowed in the list.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    Args:
+      destination: Full path to the file (ie. /bucket/object).
+      file_list: list of files to compose, see compose for details.
+      files_metadata: Meta details for the file
```
files_metadata: Metadata per file.

Missing '.'

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    Validates the compose list and builds and merges the file_list, files_metadata.
+    Args:
+      destination: Full path to the file (ie. /bucket/object).
+      file_list: list of files to compose, see compose for details.
```
s/list/List/

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  """
+    Validates the compose list and builds and merges the file_list, files_metadata.
+    Args:
+      destination: Full path to the file (ie. /bucket/object).
```
s/destination/destination_file/

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+
+def _validate_compose_list(destination_file, file_list, files_metadata=None, number_of_files=32):
+  """
+    Validates the compose list and builds and merges the file_list, files_metadata.
```
"""Validates the compose list and builds and merges the file_list, files_metadata.
 
Args:
  ....

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+      file_list: list of files to compose, see compose for details.
+      files_metadata: Meta details for the file
+      number_of_files=32: maximum number of files allowed in the list.
+    Returns:
```
A (list_of_files, bucket) tuple. list_of_files is a ....

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+                                      files_metadata, 32)
+  compose_object(file_list, destination_file, content_type)
+
+def _compose_batch(list_of_files, destination_file, files_metadata=None,
```
Here and in other places. We use 2 empty line to separate module-level methods/classes (and 1 to separate between class methods).

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+      bucket: bucket name extracted from the file paths.
+  """
+  common.validate_file_path(destination_file)
+  bucket = '/' + destination_file.split('/')[1] + '/'
```
alternative: bucket = a[0:destination_file.index('/', 1) + 1]

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  """
+  common.validate_file_path(destination_file)
+  bucket = '/' + destination_file.split('/')[1] + '/'
+  if not isinstance(file_list, list):
```
This means passing a tuple, set or other type of collection will not fail. 
I am not sure if such a check is necessary (especially when you are only using it for len and iteration and copying it
to another internal data structure).

rbruyere replied:
I could remove it. I wanted to make sure there was a meaningful error if they sent the wrong type. I just tested and both tuples and sets do cause the error to be raised. If there are are other types that might be an issue I can test them as well.

aozarov replied:
On Mon, Mar 30, 2015 at 7:41 AM, rbruyere <notifications@github.com> wrote:

> In python/src/cloudstorage/cloudstorage_api.py
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6#discussion_r27395702>
> :
>
> > +
> > +def _validate_compose_list(destination_file, file_list, files_metadata=None, number_of_files=32):
> > +  """
> > +    Validates the compose list and builds and merges the file_list, files_metadata.
> > +    Args:
> > +      destination: Full path to the file (ie. /bucket/object).
> > +      file_list: list of files to compose, see compose for details.
> > +      files_metadata: Meta details for the file
> > +      number_of_files=32: maximum number of files allowed in the list.
> > +    Returns:
> > +      list_of_files: Ready to use dict version of the list.
> > +      bucket: bucket name extracted from the file paths.
> > +  """
> > +  common.validate_file_path(destination_file)
> > +  bucket = '/' + destination_file.split('/')[1] + '/'
> > +  if not isinstance(file_list, list):
>
> I could remove it. I wanted to make sure there was a meaningful error if
> they sent the wrong type. I just tested and both tuples and sets do cause
> the error to be raised. If there are are other types that might be an issue
> I can test them as well.
>
The thing is, I don't think providing tuples or sets should  err and
therefore I think you should either remove the test or check for what you
expect (__len__ and __iter__). My preference would be the former.

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6/files#r27395702>
> .
>

aozarov commented on python/test/compos_test.py:
```diff
@@ -0,0 +1,919 @@
+# Copyright 2012 Google Inc. All Rights Reserved.
```
compose_test.py

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  list_of_files = []
+  for source_file, meta_data in itertools.izip_longest(file_list, files_metadata):
+    if not isinstance(source_file, str):
+      raise TypeError('Each item of file_list must be a string')
```
That would be a little strange error message when len(files_metadata) > len(file_list)

rbruyere replied:
Agreed, I'll add a check for that.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  for source_file, meta_data in itertools.izip_longest(file_list, files_metadata):
+    if not isinstance(source_file, str):
+      raise TypeError('Each item of file_list must be a string')
+    if source_file.startswith("/"):
```
'/'

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  Args:
+    list_of_files: List of file name strings with no leading slashes or bucket.
+    destination_file: Path to the desired output file. Must have the bucket in the path.
+    files_metadata: Optional file metadata, see link for available options:
```
Make it clear that order must match list_of_files.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,184 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+
+    list_entry = {}
+    if meta_data is not None:
+      for meta in meta_data:
```
list_entry.update(meta_data)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    be built from other existing composites, provided that the total
+    component count does not exceed 1024. See here for details:
+    https://cloud.google.com/storage/docs/composite-objects
+  Args:
```
Line separator

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+            with open(bucket + source_file['Name'], 'r') as gcs_source:
+              gcs_merge.write(gcs_source.read())
+          except errors.NotFoundError as error:
+            raise error
```
As commented before, I think this is the right thing to do (propagate the exception and log is unnecessary) but in that case any value to the try ~ except?

rbruyere replied:
I just didn't think to remove the try catch. There is no need for it now as the error is popped up.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  compose_object(file_list, destination_file, content_type)
+
+
+def _compose_batch(list_of_files, destination_file, files_metadata=None,
```
Such logic should move to the caller (e.g. MR) which can better handle its temporary files (location, name, cleanup) and chunks strategy.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    True if the file is accessible otherwise False.
+  """
+  try:
+    with open(destination, "r") as _:
```
with open(destination, "r"):
  return True

aozarov replied:
any reason you keep the " as _"?

rbruyere replied:
Chalk that up to me just seeing the return True.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    file_list: List of files to compose, see compose for details.
+    files_metadata: Meta details for the file.
+    number_of_files: Maximum number of files allowed in the list.
+  Returns:
```
line separator

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  Args:
+    destination: Full path to the file (ie. /destination_bucket/destination_file).
+    file_list: List of files to compose, see compose for details.
+    files_metadata: Meta details for the file.
```
per file (or for file_list)

aozarov replied:
Unchanged.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  """
+  common.validate_file_path(destination_file)
+  bucket = destination_file[0:(destination_file.index('/', 1) + 1)]
+  if not isinstance(file_list, list):
```
unnecessary (will cause other "valid" types such as tuples, sets to be rejected)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +277,193 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    raise ValueError("files_metadata contains more entries(%i) than file_list(%i)"
+                     % (len(files_metadata), list_len))
+
+  list_of_files = []
```
Move down near "list_entry = {}"

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,42 @@ def get_bucket_async(self, path, **kwds):
+
+    Taking a list of files it calls the put object with the compose flag
+      this call merges all the files into the destination file
+    Currently stubbed because the dev server cloudstorage_stub.py
```
Indent (and change case). or period sentence before (I find it strange to start a sentence with "Currently")...
Also, is that sentence relevant to this file (and method)?

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,42 @@ def get_bucket_async(self, path, **kwds):
+      file_list: file_list: list of dicts with the file name.
+      destination_file: Path to the destination file.
+      content_type: Content type for the destination file.
+      retry_params: An api_utils.RetryParams for this call to GCS. If None,
```
wrong copy/paste (here and bellow).

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,42 @@ def get_bucket_async(self, path, **kwds):
+    """
+    xml = ''
+    for meta_data in file_list:
+      xml_meta = ""
```
be consistent (and use style in file ) about " or ' (in this case should be ')

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,42 @@ def get_bucket_async(self, path, **kwds):
+      for meta in meta_data:
+        xml_meta += "<%s>%s</%s>" % (meta, meta_data[meta], meta)
+      xml += '<Component>%s</Component>' % xml_meta
+    xml = '<ComposeRequest>%s</ComposeRequest>' % xml
```
That is not an efficient way to build the string

rbruyere replied:
How is it not efficient. Any recommendations?

aozarov replied:
On Mon, Mar 30, 2015 at 1:44 PM, rbruyere <notifications@github.com> wrote:

> In python/src/cloudstorage/storage_api.py
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6#discussion_r27429987>
> :
>
> > +
> > +    Args:
> > +      file_list: file_list: list of dicts with the file name.
> > +      destination_file: Path to the destination file.
> > +      content_type: Content type for the destination file.
> > +      retry_params: An api_utils.RetryParams for this call to GCS. If None,
> > +      the default one is used.
> > +    _account_id: Internal-use only.
> > +    """
> > +    xml = ''
> > +    for meta_data in file_list:
> > +      xml_meta = ""
> > +      for meta in meta_data:
> > +        xml_meta += "<%s>%s</%s>" % (meta, meta_data[meta], meta)
> > +      xml += '<Component>%s</Component>' % xml_meta
> > +    xml = '<ComposeRequest>%s</ComposeRequest>' % xml
>
> How is it not efficient. Any recommendations?
>
In general a + b + c would be better than (a + "%s" + c) % b  (think from
allocated bytes and byte iteration perspective).
So, in your case plain initialization xml_meta with "<Component>" and
adding "</<Component>" after iteration and doing
the same for xml would be way more efficient (and without doing anything
special).
If you really want to be efficient you should consider (cStringIO or join)
- see:  https://waymoot.org/home/python_string/

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/appengine-gcs-client/pull/6/files#r27429987>
> .
>

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+  """
+  return [ int(char) if char.isdigit() else char for char in re.split('([0-9]+)', in_string) ]
+# pylint: disable=too-few-public-methods
+class MainPage(webapp2.RequestHandler):
```
2 line seperator

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+  """ Turn a string into a list of string and number chunks.
+      "z23a" -> ["z", 23, "a"]
+  """
+  return [ int(char) if char.isdigit() else char for char in re.split('([0-9]+)', in_string) ]
```
remove the space after "[" and before "]"

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+  Used to test the cloudstorage compose method
+  test_sizes_to_run is used to specify how many files to send.
+  """
+  # pylint: disable=too-many-locals, too-many-statements
```
line separator

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+import logging
+import re
+import cloudstorage
+from google.appengine.api import app_identity
```
Lets get Johns input on that as well but I fear that you many need to maintain the package hackery seen in other tests.

aozarov replied:
Did you speak with John. I think this may still be an issue and would need to change to:

try:
  from cloudstorage import errors
except ImportError:
  from google.appengine.ext.cloudstorage import errors

Like in cloudstorage_test.py

lucena replied:
I believe the complete set of imports should be provided as follows:

```python
import logging
import re
import webapp2

from google.appengine.api import app_identity

try:
  import cloudstorage
  from cloudstorage import errors
except ImportError:
  from google.appengine.ext import cloudstorage
  from google.appengine.ext.cloudstorage import errors
```

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+    self.response.out.write(output)
+    self.response.out.write("</BODY></HTML>")
+# pylint: disable=too-many-arguments
+def do_test(test_sizes_to_run, template,
```
I don't see why this should not follow the same testing style as other tests (and use unittest)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +276,130 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+
+  Args:
+    list_of_files: List of file name strings with no leading slashes or bucket.
+    destination_file: Path to the desired output file. Must have the bucket in the path.
```
Here and in several other places you are exceeding 80 columns.

rbruyere replied:
Oddly pylint had 100 characters set as the default line length. Fixing all of them now

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +276,130 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  if files_metadata is None:
+    files_metadata = []
+
+  if len(files_metadata) > list_len:
```
Nit, if you care you can put that as part of an "else" above.

rbruyere replied:
Sorry I don't understand/

aozarov replied:
In this case, because your conditions are raising exception upon true it is just a matter of style,
therefore feel free to ignore (if list_len > number_of_files: ... elif: list_len <= 1 ...)

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +276,130 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+  if len(files_metadata) > list_len:
+    raise ValueError("files_metadata contains more entries(%i) than file_list(%i)"
+                     % (len(files_metadata), list_len))
+  list_of_files = []
```
I suggested before to move it down (closer to where it is first used).

rbruyere replied:
Needs to be above the For loop so it can be referenced after (in the return)

aozarov replied:
Yes, my mistake.

aozarov commented on python/src/cloudstorage/cloudstorage_api.py:
```diff
@@ -273,6 +276,130 @@ def listbucket(path_prefix, marker=None, prefix=None, max_keys=None,
+    if source_file.startswith(bucket):
+      logging.warn('Detected bucket name at the start of the file, '
+                   'must not specify the bucket when listing file_names.'
+                   ' May cause files to be miss read')
```
misread ?

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,43 @@ def get_bucket_async(self, path, **kwds):
+  def compose_object(self, file_list, destination_file, content_type):
+    """COMPOSE multiple objects together.
+
+    Taking a list of files it calls the put object with the compose flag
```
Alternative: Using the given list of files calls the put object with the compose flag.

and decide about a comma or a dot and Uppercase bellow.

aozarov commented on python/src/cloudstorage/storage_api.py:
```diff
@@ -169,6 +169,43 @@ def get_bucket_async(self, path, **kwds):
+      headers = None
+    # pylint: disable=no-member
+    status, resp_headers, content = self.put_object(
+                                       api_utils._quote_filename(destination_file) + '?compose',
```
still indent issues

aozarov commented on python/src/cloudstorage/.settings/de.loskutov.FileSync.prefs:
```diff
@@ -0,0 +1,7 @@
+WARNING=DO NOT MODIFY THIS FILE IF YOU DON'T UNDERSTAND
```
We should not include this file but rather add a .settings to .gitignore

aozarov replied:
I still see this file...

rbruyere replied:
It was in the git ignore. I had to manually delete it before the update would remove it. Its gone now.

aozarov commented on python/test/compose_test.py:
```diff
@@ -0,0 +1,99 @@
+import cloudstorage
+from google.appengine.api import app_identity
+from cloudstorage import errors
+
```
This is still not a unit test.

aozarov commented on .gitignore:
```diff
@@ -0,0 +1,6 @@
+.project
+.pydevproject
+.settings
+/python/test/cloudstorage
```
This line does not seem necessary

rbruyere replied:
I had the folder in there to test the unit test. Didn't think that should be uploaded. Ill upload a cleaner version

aozarov commented on python/test/cloudstorage_test.py:
```diff
@@ -504,5 +506,67 @@ def AssertListBucketEqual(self, expected, bucket):
+
+  def testComposeInvalidItemInFileListFailure(self):
+    """Test to ensure ValueError is thrown if there are invalid entries"""
+    self.assertRaises(TypeError, cloudstorage.compose, ["1", 1], DESTFILE)
```
'1'

aozarov commented:
Ok, looks good. Let me know if you also think it is ready to be merged.
